### ISSUE TITLE
lockscreen playback issue

### ISSUE THREAD

**Issue by wolfensg** [droidmjt/Droidsound#492]
It appears that on stock Nougat, Samsung Galaxy S6 DroidSound is stuttering when device is locked after a while (i.e. it plays everything as it should but with some halfsecond breaks every couple of seconds).

I have also seen similar behaviour when using A3 2016 with unlocked second cluster of CPU cores (disabled by Samsung due to battery limitations,etc so oficialy device is quadcore) but did not report then as I just don't want to report anything running on modified kernel and some nonstock featues.

Reproduceable 100% of times.

Enabling partial or full wakeup in DroidSound settings does not seem to help.
Any thoughts? or is this a bug?

**Comment by droidmjt** [comment]
with what format this happens?

**Comment by wolfensg** [comment]
Actually maybe thats a good question. I've tested sap (Atari), sid (C64), couple of modules (XM) and MP3s and:

sap: plays properly
sid: plays properly but stutters on lockscreen (when device locked/screen turned off)
various mod files: white noise generated instead of playback
various mp3 files: white noise as well

it appears that I'm also on previous build (167) I'll uninstall, clean cache and install latest and restes

**Comment by wolfensg** [comment]
Ok retested with build 168.
All issues gone :)

**Comment by wolfensg** [comment]
Ok. It still happens from time to time - for sure with c64 sids. I'll retest with couple of more formats (fortunatelly this whitenoise bug was only an issue on build 167)

**Comment by wolfensg** [comment]
With 168 build I wasn't able to reproduce issue with any of the MP3s, Modules (XM,IT,S3M) oraz with Atari SAP format songs. In fact it appears it DID not EVER happened with any of those formats.
Issue does seem however to occur with C64 sids. Steps to reproduce:
1. Start playback of any song longer than 1 minute (well... it doesn't make any sense to test with shorter ones ;)
2. Lock the phone. (and wait)
3. Issue appears more or less randomly after couple of seconds (may take up to 1 minute to occur). Sometimes it is possible to play the whole song without interruption.
Issue description: Gap in the playback with silence for half a second (pausing playback) or so. Sometimes this "gap" is repeated couple of times with the pattern: song played, pause, song resumes to play, pause, song again resumes, etc. (sometimes (but I may have overheard it) it seems that the gaps are increasingly (by some small percentage) longer if happening couple of times in the row.

**Comment by droidmjt** [comment]
Tested with SGS6 and it happens during the sleep mode (screen is turned off). I'll recheck if the wakelock is working correctly, it doesn't happen with more powerful phones tho.

**Comment by droidmjt** [comment]
https://developer.android.com/about/versions/nougat/android-7.0-changes
I was reading about the Doze mode and it seems to be the curse of Android Nougat. Several pages claims the Doze mode can override the wakelocks, I have this shitty feeling there is not much I can do here :(

**Comment by wolfensg** [comment]
The odd thing is that my second phone (mentioned already A3 2016) is also running Nougat 7.0 and does not have this issue. However when I enabled second cluster of CPU cores (normally disabled) issue started appearing. When I flashed the standard kernel with only one cluster of 4 cores enabled all again was fine.
S6 has by default 2 clusters - one speced higher than the second.
Maybe it is somehow related to the CPU load handling as well (and offloading something from cluster which is being forced to sleep)?

**Comment by droidmjt** [comment]
makes sense, but I'm wondering, with one cluster, are those 4 cores the big or the LITTLE ones, see link below

https://en.wikipedia.org/wiki/ARM_big.LITTLE

**Comment by wolfensg** [comment]
I am testing now small workaround (probably costing some battery but hell... it's worth it:)

In SGS6 settings / device maintenance / battery here two things to do:
1. Make sure DroidSound is in "Unmonitored Apps" list
2. Click "Battery Usage" then three dot menu then "Optimise battery usage" select from dropdown "All apps". Find DroidSound on the list and disable optimise battery usage.

**Comment by wolfensg** [comment]
On A3 2016 both clusters are little (Cortex-A53). But on S6 there is a difference.

**Comment by droidmjt** [comment]
I tested this with SGS3 (using Resurrection ROM Android 7.1.2), with default settings it works normally.
The CPU is just 1.4GHz, 4 cores, no idea how many clusters.

**Comment by wolfensg** [comment]
I tested also on A40 and also works fine but requires the above 'fix' with disabling battery optimisations (which do not work on S6 unfortunately). A40 has 2 clusters, one with 2 cores, second with 6 cores clocked slightly lower.

**Comment by droidmjt** [comment]
Since SGS3 with 1.4GHz (Cortex-A9) is able to play even 2SIDs during sleep mode, something is wrong with the powersaving in SGS6 or Samsung version of Nougat.

Note, playing 2SID files require twice the juice BTW. Not to mention 3SIDs.

**Comment by droidmjt** [comment]
Tested now with Samsung Galaxy S Duos 2 Trend S7582, stock ROM, Dual-core CPU (Cortex-A9) and same SID works while the screen is turned off, no powersaving optimizations applied. It seems the Cortex-A9 was a decent CPU.

**Comment by wolfensg** [comment]
Yup. Works also on Galaxy Ace2 (dualcore NovaThor clocked at 800Mhz)

**Comment by droidmjt** [comment]
I'll close this, nothing to fix here

**Comment by wolfensg** [comment]
Just a note: ZXTune and MoDo players work fine when the sleep mode kicks in.

**Comment by droidmjt** [comment]
MoDo uses lighter library to relpay SID files. ZXTune works only when Interpolation is None, you can achieve the same in DSE by enabling Faster Emulation option from settings, works only with ReSID.

**Comment by wolfensg** [comment]
Thanks. Will test this in some longer free time:)

**Comment by wolfensg** [comment]
Yup. You were right with the faster emulation setting (and interpolation in ZXTune which I had disabled). Although SID playback is worse now at least there is no problem during doze. Looks like this has to do as a workaround for now. Thanks!:)

**Comment by wolfensg** [comment]
Since the need to have top quality audio (which with DroidSound-E is totally possible) is really important for me and no other players provides support for so many music formats - I've tried to experiment with some other tweaks.

Soo... I've installed TWRP and rooted my Galaxy S6 (with Magisk) and installed custom kernel (Aldeon Kernel -> https://forum.xda-developers.com/galaxy-s6/development/7-0-samsung-galaxy-s6-aldeon-kernel-t3617819 ). When I've set governors to SmartAssV2 (for both clusters) and bumped minimum clock rate for both clusters from 400Mhz to 500Mhz all seems to be stable (no playback glitches, etc).
Probably it would be possible to make the setup more battery friendly but at least everything works properly now and cost is minimal (hardly noticeable less total time on battery).

Anyway - @droidmjt - Thanks for really good hints!
